lib/bpg.ts

```diff
@@ -56,14 +80,20 @@ function generateKey(object) {
 }
 
 class BPGObject {
-  constructor(t, name) {
+  _typeName: string;
+  _type: FieldType;
+  fields: { [key: string]: any };
+  _parent?: BPGObject;
+  _key?: { [key: string]: any };
+
+  constructor(t: FieldType, name: string) {
     this._typeName = name;
     this._type = t;
     this.fields = {};
   }
```

Comment:
You can use instead:

```suggestion
  private readonly fields: Record<string, unknown> = {}
  _parent?: BPGObject;
  _key?: Record<string, any>;

  constructor(private readonly _type: FieldType, private readonly _typeName: string) {
}
```

Reply:
There are like a million places here that could be written in a more TypeScript way, but I don't think this should be the focus of this PR.